asset/asset.go

```diff
@@ -801,20 +798,105 @@ var _ GenesisSigner = (*RawKeyGenesisSigner)(nil)
 
 // DeriveGroupKey derives an asset's group key based on an internal public
 // key descriptor, the original group asset genesis, and the asset's genesis.
-func DeriveGroupKey(genSigner GenesisSigner, rawKey keychain.KeyDescriptor,
-	initialGen Genesis, currentGen *Genesis) (*GroupKey, error) {
+func DeriveGroupKey(genSigner GenesisSigner, groupKey *GroupKey,
+	scriptLeaf *psbt.TaprootTapLeafScript, initialGen Genesis,
+	newAsset *Asset) (*GroupKey, error) {
+
+	// First, perform the final checks on the asset being authorized for
+	// group membership.
+	if newAsset == nil || newAsset.IsUnknownVersion() {
+		return nil, fmt.Errorf("genesis asset is missing or unknown " +
+			"version")
+	}
+
+	if newAsset.Genesis != initialGen && initialGen.Type != newAsset.Type {
+		return nil, fmt.Errorf("asset group type mismatch")
+	}
+
+	if scriptLeaf != nil && groupKey.TapscriptRoot == [32]byte{} {
```

Comment:
At this phase, I don't think we can really assert much about the script leaf. 

For now we can handle just the keyspend path. Anything more advanced than that (musig2, etc, etc) will require the caller to drive the API to do the fancy stuff.